### PR TITLE
[FIX] visibility errors on gcc - range

### PR DESCRIPTION
**Pull request by marehr**
Superseeds #1228

fixes partially #1317

### CODE REVIEW

File: include/seqan3/range/container/dynamic_bitset.hpp [hunk 1]
```diff
@@ -1580,62 +1587,44 @@ class dynamic_bitset
      */
     //!\brief Performs element-wise comparison.
     template <size_t cap>
-    //!\cond
-        requires cap <= bit_capacity /* resolves ambiguousness when comparing two dynamic_bitsets of unequal capacity */
-    //!\endcond
-    friend constexpr bool operator==(dynamic_bitset const & lhs, dynamic_bitset<cap> const & rhs) noexcept
+    constexpr bool operator==(dynamic_bitset<cap> const & rhs) const noexcept
```
Comment (smehringer) [inline include/seqan3/range/container/dynamic_bitset.hpp]: why this change? I thought we always declare comprison operators as friends?

File: include/seqan3/range/container/small_string.hpp [hunk 2]
```diff
@@ -158,11 +158,12 @@ class small_string : public small_vector<char, capacity_ + 1>
      *
      * No-throw guarantee if value_type is std::is_nothrow_copy_constructible.
      */
-    template <std::forward_iterator begin_it_type, std::sentinel_for<begin_it_type> end_it_type>
-    constexpr void assign(begin_it_type begin_it, end_it_type end_it) noexcept
+    template <std::forward_iterator begin_it_type, typename end_it_type>
     //!\cond
-        requires std::constructible_from<value_type, /*ranges::iter_reference_t*/reference_t<begin_it_type>>
+        requires std::sentinel_for<end_it_type, begin_it_type> &&
+                 std::constructible_from<value_type, /*ranges::iter_reference_t*/reference_t<begin_it_type>>
```
Comment (smehringer) [inline include/seqan3/range/container/small_string.hpp]:
```suggestion
                 std::constructible_from<value_type, std::ranges::iter_reference_t<begin_it_type>>
```
I think this should work now?

File: include/seqan3/range/views/pairwise_combine.hpp [hunk 3]
```diff
@@ -130,8 +130,9 @@ class pairwise_combine_view : public std::ranges::view_interface<pairwise_combin
          * Allows construction of a const iterator (operating on a const range) from a non-const iterator.
          * This special constructor is needed as it is not covered by the standard copy and move constructors.
          */
-        template <std::convertible_to<range_type &> other_range_type>
-            requires std::same_as<std::remove_const_t<other_range_type>, std::remove_const_t<range_type>>
+        template <typename other_range_type>
+            requires std::convertible_to<other_range_type, range_type &> &&
```
Comment (smehringer) [inline include/seqan3/range/views/pairwise_combine.hpp]: doxugen `//!\cond`

File: include/seqan3/range/container/small_string.hpp (hunk 2 above)
Comment (marehr) [reply include/seqan3/range/container/small_string.hpp]: thank you for the catch, I look into it.

File: include/seqan3/range/container/dynamic_bitset.hpp (hunk 1 above)
Comment (marehr) [reply include/seqan3/range/container/dynamic_bitset.hpp]: I try to remember, but there was a problem. Enrico wrote: resolves ambiguousness when comparing two dynamic_bitsets of unequal capacity. I think this is a different issue.

I think the problem was that

```c++
friend constexpr bool operator==(dynamic_bitset const & lhs, dynamic_bitset<cap> const & rhs) noexcept;
```

means: my current class instance `dynamic_bitset` befriends `operator==` where first argument is my current class instance and the second argument is any other class instance of dynamic_bitset. (NOTE: they don't befriend this function overload)

This means that `operator==` is only allowed to access `lhs`, but not `rhs`, because `rhs` is not necessary the same type as the befriending `lhs`.

TLDR; Only lhs type befriends that class, rhs type does not.

----

I'm not to sure if we should always use `friend` declarations. 

I read https://blogs.accu.org/?p=1974 and then tried to play around with https://godbolt.org/z/ZmIqxA.

AFAIU we befriend to have the `hidden friend` property that means that a friend declared function WILL not be found by ADL of other types. And it allows to implicitly converts the left hand side in your type (if you have an implicit conversion constructor).

That means we lose in this case the implicit conversion of the left hand side. If it's important we need to figure out something else.

If we want this, we need to declare
```c++
    template <size_t cap1, size_t cap2>
    friend inline constexpr bool operator==(dynamic_bitset<cap1> const & lhs, dynamic_bitset<cap2> const & rhs) noexcept;
```
and define the friend function outside once, but then any implicit conversation from any other type to bitset would use that comparison operator. 

Making the private members protected would work.

Comment (marehr) [reply include/seqan3/range/container/dynamic_bitset.hpp]: Maybe the more killer argument:

```c++
#include <bitset>

int main()
{
    std::bitset<5> a{};
    a == 5; // works
    // 5 == a; // does not works (because operator== is a member)
    return 0;
}
```

Comment (smehringer) [reply include/seqan3/range/container/dynamic_bitset.hpp]: Phew I am not sure if I understood everything :see_no_evil: 
let's talk about this in person

Comment (marehr) [reply include/seqan3/range/container/dynamic_bitset.hpp]: yeah, I talked with @rrahn and he was in favour for the member instead of friend. If I understood @eseiler correctly, than he doesn't mind. Maybe we should put it as a discussion for next monday.

File: include/seqan3/range/container/small_string.hpp (hunk 2 above)
Comment (marehr) [reply include/seqan3/range/container/small_string.hpp]: I just saw that I didn't comment this out, so this is not in the scope of my PR.

File: include/seqan3/range/container/dynamic_bitset.hpp (hunk 1 above)
Comment (marehr) [reply include/seqan3/range/container/dynamic_bitset.hpp]: We know use the public api.

File: include/seqan3/range/container/dynamic_bitset.hpp [hunk 4]
```diff
@@ -1082,13 +1088,13 @@ class dynamic_bitset
     //!\brief Direct access to the underlying bit field.
     constexpr bitfield * raw_data() noexcept
     {
-        return data;
+        return &data;
```
Comment (smehringer) [inline include/seqan3/range/container/dynamic_bitset.hpp]: how was this not caught before? ...

File: include/seqan3/range/container/dynamic_bitset.hpp [hunk 5]
```diff
@@ -1580,59 +1587,41 @@ class dynamic_bitset
      */
     //!\brief Performs element-wise comparison.
     template <size_t cap>
-    //!\cond
-        requires cap <= bit_capacity /* resolves ambiguousness when comparing two dynamic_bitsets of unequal capacity */
```
Comment (smehringer) [inline include/seqan3/range/container/dynamic_bitset.hpp]: (btw. you say the outcommenting std::ranges::iter_reference_t in line 1289 is not in the scope of this PR but here you also change something unrelated?)

Comment (marehr) [reply include/seqan3/range/container/dynamic_bitset.hpp]: Yep, because this is self-contained, but this change would be needed across multiple files.

Comment (marehr) [reply include/seqan3/range/container/dynamic_bitset.hpp]: If you want, I can change the requires back.

File: include/seqan3/range/container/dynamic_bitset.hpp (hunk 4 above)
Comment (marehr) [reply include/seqan3/range/container/dynamic_bitset.hpp]: Because it is a constexpr (implicitly inline) function and thus will not be "instantiated" The coverage says also 100% even though it isn't, because all those constexpr functions are hidden.

Comment (smehringer) [reply include/seqan3/range/container/dynamic_bitset.hpp]: How to trick the coverage :smile:

File: include/seqan3/range/container/dynamic_bitset.hpp (hunk 5 above)
Comment (smehringer) [reply include/seqan3/range/container/dynamic_bitset.hpp]: no it's fine. just fix travis and I'll merge

**Comment by marehr** [comment]
@smehringer Travis is failing because this PR is depending on #1307. I'll ping here once the other one was merged.

**Comment by marehr** [comment]
@smehringer Merge me :o)
